File: src/surface-generation/biomes/climate.ts
import { Pair } from "./consumer"

export interface Sampler {
    sample(x: number, y: number, z: number): TargetPoint
}

export class Parameter {
    constructor(public min: number, public max: number) {}

    static span(min: Parameter, max: Parameter): Parameter
    static span(min: number, max: number): Parameter
    static span(min: number | Parameter, max: number | Parameter): Parameter {
        if (typeof min === "number") {
            if (typeof max !== "number") {
                throw new Error("")
            }
            return new Parameter(quantizeCoord(min), quantizeCoord(max))
        } else {
            if (typeof max === "number") {
                throw new Error("")
            }
            return new Parameter(min.min, max.max)
        }
    }

    static point(value: number): Parameter {
        return Parameter.span(value, value)
    }

    distance(value: number): number {
        const v0 = value - this.max
        const v1 = this.min - value
        return v0 > 0 ? v0 : Math.max(v1, 0)
    }

    get center(): number {
        return (this.min + this.max) * 0.5
    }

    get length(): number {
        return this.max - this.min
    }
}

export class TargetPoint {
    constructor(
        public temperature: number,
        public humidity: number,
        public continentalness: number,
        public erosion: number,
        public depth: number,
        public weirdness: number
    ) {}
}

export class ParameterPoint {
    constructor(
        public temperature: Parameter,
        public humidity: Parameter,
        public continentalness: Parameter,
        public erosion: Parameter,
        public depth: Parameter,
        public weirdness: Parameter,
        public offset: number
    ) {}

    fitness(targetPoint: TargetPoint): number {
        const temperatureDistance = this.temperature.distance(targetPoint.temperature)
        const humidityDistance = this.humidity.distance(targetPoint.humidity)
        const continentalnessDistance = this.continentalness.distance(targetPoint.continentalness)
        const erosionDistance = this.erosion.distance(targetPoint.erosion)
        const depthDistance = this.depth.distance(targetPoint.depth)
        const weirdnessDistance = this.weirdness.distance(targetPoint.weirdness)
        const offsetDistance = 0 - this.offset

        return (
            temperatureDistance * temperatureDistance +
            humidityDistance * humidityDistance +
            continentalnessDistance * continentalnessDistance +
            erosionDistance * erosionDistance +
            depthDistance * depthDistance +
            weirdnessDistance * weirdnessDistance +
            offsetDistance * offsetDistance
        )
    }
}

const QUANTIZATION_FACTOR = 10000

export function quantizeCoord(coord: number): number {
    return Math.trunc(coord * QUANTIZATION_FACTOR)
}

export function unquantizeCoord(coord: number): number {
    return coord / QUANTIZATION_FACTOR
}

export function parameters(
    temperature: number,
    humidity: number,
    continentalness: number,
    erosion: number,
    depth: number,
    weirdness: number,
    offset: number
): ParameterPoint

export function parameters(
    temperature: Parameter | number,
    humidity: Parameter | number,
    continentalness: Parameter | number,
    erosion: Parameter | number,
    depth: Parameter | number,
    weirdness: Parameter | number,
    offset: number
): ParameterPoint

export function parameters(
    temperature: Parameter | number,
    humidity: Parameter | number,
    continentalness: Parameter | number,
    erosion: Parameter | number,
    depth: Parameter | number,
    weirdness: Parameter | number,
    offset: number
): ParameterPoint {
    if (typeof temperature === "number") {
        if (
            typeof temperature !== "number" ||
            typeof humidity !== "number" ||
            typeof continentalness !== "number" ||
            typeof erosion !== "number" ||
            typeof depth !== "number" ||
            typeof weirdness !== "number"
        ) {
            throw new Error("")
        }

        return new ParameterPoint(
            Parameter.point(temperature),
            Parameter.point(humidity),
            Parameter.point(continentalness),
            Parameter.point(erosion),
            Parameter.point(depth),
            Parameter.point(weirdness),
            quantizeCoord(offset)
        )
    } else {
        if (
            typeof temperature === "number" ||
            typeof humidity === "number" ||
            typeof continentalness === "number" ||
            typeof erosion === "number" ||
            typeof depth === "number" ||
            typeof weirdness === "number"
        ) {
            throw new Error("")
        }

        return new ParameterPoint(
            temperature,
            humidity,
            continentalness,
            erosion,
            depth,
            weirdness,
            quantizeCoord(offset)
        )
    }
}

export function findValueBruteForce<T>(
    targetPoint: TargetPoint,
    values: Pair<ParameterPoint, T>[]
): T {
    let minDistance = Infinity
    let result: T | undefined

    for (const pair of values) {
        const distance = pair.first.fitness(targetPoint)
        if (distance < minDistance) {
            minDistance = distance
            result = pair.second
        }
    }

    if (result === undefined) {
        throw new Error("Result not found.")
    }

    return result
}

export function target(
    temperature: number,
    humidity: number,
    continentalness: number,
    erosion: number,
    depth: number,
    weirdness: number
): TargetPoint {
    return new TargetPoint(
        quantizeCoord(temperature),
        quantizeCoord(humidity),
        quantizeCoord(continentalness),
        quantizeCoord(erosion),
        quantizeCoord(depth),
        quantizeCoord(weirdness)
    )
}

File: src/cpp/surface-generation/biomes/climate.cpp
#include "climate.hpp"

// TargetPoint

Climate::TargetPoint::TargetPoint(int64_t temperature, int64_t humidity, int64_t continentalness, int64_t erosion,
                                  int64_t depth, int64_t weirdness) {
    this->temperature = temperature;
    this->humidity = humidity;
    this->continentalness = continentalness;
    this->erosion = erosion;
    this->depth = depth;
    this->weirdness = weirdness;
}

// Parameter

Climate::Parameter::Parameter(int64_t min, int64_t max) : min(min), max(max) {
}

Climate::Parameter const Climate::Parameter::point(float value) {
    return span(value, value);
}

Climate::Parameter const Climate::Parameter::span(float min, float max) {
    return Climate::Parameter(Climate::quantizeCoord(min), Climate::quantizeCoord(max));
}

Climate::Parameter const Climate::Parameter::span(Climate::Parameter const &minParameter,
                                                  Climate::Parameter const &maxParameter) {
    return Climate::Parameter(minParameter.min, maxParameter.max);
}

int64_t Climate::Parameter::distance(int64_t value) const {
    int64_t distanceFromMax = value - this->max;
    int64_t distanceFromMin = this->min - value;
    return distanceFromMax > 0LL ? distanceFromMax : std::max(distanceFromMin, 0LL);
}

int64_t Climate::Parameter::distance(Climate::Parameter const &parameter) const {
    int64_t distanceFromMax = parameter.min - this->max;
    int64_t distanceFromMin = this->min - parameter.max;
    return distanceFromMax > 0LL ? distanceFromMax : std::max(distanceFromMin, 0LL);
}

Climate::Parameter const Climate::Parameter::span(Climate::Parameter const &parameterToMerge) const {
    return parameterToMerge.isNull() ? *this : Climate::Parameter(std::min(this->min, parameterToMerge.min), std::max(this->max, parameterToMerge.max));
}

bool Climate::Parameter::isNull() const {
    return this->min > this->max;
}

// ParameterPoint

Climate::ParameterPoint::ParameterPoint(Climate::Parameter const &temperature, Climate::Parameter const &humidity,
                                        Climate::Parameter const &continentalness, Climate::Parameter const &erosion,
                                        Climate::Parameter const &depth, Climate::Parameter const &weirdness,
                                        int64_t offset)
    : temperature(temperature), humidity(humidity), continentalness(continentalness), erosion(erosion), depth(depth),
      weirdness(weirdness), offset(offset) {
}

int64_t Climate::ParameterPoint::fitness(Climate::TargetPoint const &otherPoint) const {
    return Mth::square(this->temperature.distance(otherPoint.temperature)) +
           Mth::square(this->humidity.distance(otherPoint.humidity)) +
           Mth::square(this->continentalness.distance(otherPoint.continentalness)) +
           Mth::square(this->erosion.distance(otherPoint.erosion)) +
           Mth::square(this->depth.distance(otherPoint.depth)) +
           Mth::square(this->weirdness.distance(otherPoint.weirdness)) + Mth::square(this->offset);
}

// Climate

Climate::TargetPoint const Climate::target(float temperature, float humidity, float continentalness, float erosion,
                                           float depth, float weirdness) {
    return Climate::TargetPoint(quantizeCoord(temperature), quantizeCoord(humidity), quantizeCoord(continentalness),
                                quantizeCoord(erosion), quantizeCoord(depth), quantizeCoord(weirdness));
}

Climate::ParameterPoint const Climate::parameters(float temperature, float humidity, float continentalness,
                                                  float erosion, float depth, float weirdness, float offset) {
    return Climate::ParameterPoint(Climate::Parameter::point(temperature), Climate::Parameter::point(humidity),
                                   Climate::Parameter::point(continentalness), Climate::Parameter::point(erosion),
                                   Climate::Parameter::point(depth), Climate::Parameter::point(weirdness),
                                   quantizeCoord(offset));
}

Climate::ParameterPoint const Climate::parameters(Climate::Parameter const &temperature,
                                                  Climate::Parameter const &humidity,
                                                  Climate::Parameter const &continentalness,
                                                  Climate::Parameter const &erosion, Climate::Parameter const &depth,
                                                  Climate::Parameter const &weirdness, float offset) {
    return Climate::ParameterPoint(temperature, humidity, continentalness, erosion, depth, weirdness,
                                   quantizeCoord(offset));
}